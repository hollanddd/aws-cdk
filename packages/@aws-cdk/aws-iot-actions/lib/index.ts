export * from './lambda';
export * from './republish';
export * from './sns';
export * from './sqs';
