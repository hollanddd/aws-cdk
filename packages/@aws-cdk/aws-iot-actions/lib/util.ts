import * as iam from '@aws-cdk/aws-iam';
import { IConstruct, Stack } from '@aws-cdk/core';

// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
// eslint-disable-next-line no-duplicate-imports, import/order
import { Construct } from '@aws-cdk/core';

/**
 * Obtain the Role for the Topic Rule
 *
 * If a role already exists, it will be returned. This ensures that if multiple
 * events have the same target, they will share a role.
 */
export function singletonTopicRuleRole(scope: IConstruct, policyStatements: iam.PolicyStatement[]): iam.IRole {
  const stack = Stack.of(scope);
  const id = 'AllowIot';
  const existing = stack.node.tryFindChild(id) as iam.IRole;
  if (existing) { return existing; }

  const role = new iam.Role(scope as Construct, id, {
    assumedBy: new iam.ServicePrincipal('iot.amazonaws.com'),
  });

  policyStatements.forEach(role.addToPolicy.bind(role));

  return role;
}
