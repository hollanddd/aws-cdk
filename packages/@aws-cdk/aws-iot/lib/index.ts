export * from './certificate';
export * from './policy';
export * from './policy-document';
export * from './policy-statement';
export * from './thing';
export * from './topic-rule';
export * from './topic-rule-action';

// AWS::IoT CloudFormation Resources:
export * from './iot.generated';
