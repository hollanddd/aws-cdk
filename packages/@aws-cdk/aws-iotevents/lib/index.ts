export * from './detector-model';
export * from './detector-model-action';
export * from './input';
export * from './payload';

// AWS::IoTEvents CloudFormation Resources:
export * from './iotevents.generated';
