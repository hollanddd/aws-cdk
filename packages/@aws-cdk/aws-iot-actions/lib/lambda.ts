import * as iam from '@aws-cdk/aws-iam';
import * as iot from '@aws-cdk/aws-iot';
import * as lambda from '@aws-cdk/aws-lambda';
import * as cdk from '@aws-cdk/core';

/**
 * Calls an AWS Lambda function
 */
export class LambdaFunction implements iot.ITopicRuleAction {
  constructor(private readonly handler: lambda.IFunction) {
  }

  public bind(rule: iot.ITopicRule): iot.TopicRuleActionConfig {
    // Allow rule to invoke lambda function
    const permissionId = 'AllowIot';
    if (!this.handler.permissionsNode.tryFindChild(permissionId)) {
      this.handler.addPermission(permissionId, {
        action: 'lambda:InvokeFunction',
        principal: new iam.ServicePrincipal('iot.amazonaws.com'),
        sourceAccount: cdk.Aws.ACCOUNT_ID,
      });
    }

    // Ensure permission is deployed before rule
    const permission = this.handler.permissionsNode.tryFindChild(permissionId) as lambda.CfnPermission;
    if (permission) {
      rule.node.addDependency(permission);
    } else {
      // tsling:disable-next-line:max-line-length
      rule.node.addWarning('This rule is using a Lambda action with an imported function. Ensure permssion is given to IOT to invoke that function.');
    }

    return {
      lambda: {
        functionArn: this.handler.functionArn,
      },
    };
  }
}
