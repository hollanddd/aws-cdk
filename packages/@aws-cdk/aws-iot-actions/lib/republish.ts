import * as iam from '@aws-cdk/aws-iam';
import * as iot from '@aws-cdk/aws-iot';
import { Stack, Arn } from '@aws-cdk/core';
import { singletonTopicRuleRole } from './util';

/**
 * Construction properties for a Republish action.
 */
export interface RepublishProps {
  /**
   * The Quality of Service (Qos) level to use when republishing messages.
   *
   * @default - 0
   */
  readonly qos?: number;
  /**
   *
   * The MQTT topic to which to republish the message.
   *
   * To republish to a reserved topic, which begins with `$`, use `$$` instead.
   *
   * For example, to republish to the device shadow topic
   * `$aws/things/MyThing/shadow/update`, specify the topic as
   * `$$aws/things/MyThing/shadow/update`.
   */
  readonly topic: string;
  /**
   * The IAM role that grants access.
   *
   * @default - a role will be created
   */
  readonly role?: iam.IRole;
}

/**
 * Publishes to a IoT Topic
 */
export class Republish implements iot.ITopicRuleAction {
  constructor(private readonly props: RepublishProps) {
  }

  public bind(rule: iot.ITopicRule): iot.TopicRuleActionConfig {
    const stack = Stack.of(rule);
    // Allow rule to publish to topic
    // TODO: accept topic rule as prop - this.props.topic.topicRuleArn;
    // TODO: grantable topic rules - this.props.topic.grantPublish(rule);
    const role = this.props.role || singletonTopicRuleRole(rule, [new iam.PolicyStatement({
      actions: ['iot:Publish'],
      resources: [
        Arn.format({
          resource: 'topic',
          service: 'iot',
          resourceName: this.props.topic,
        }, stack),
      ],
    })]);

    return {
      republish: {
        qos: this.props.qos || 0,
        topic: this.props.topic,
        roleArn: role.roleArn,
      },
    };
  }
}
