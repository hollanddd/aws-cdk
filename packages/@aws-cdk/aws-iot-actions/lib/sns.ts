import * as iam from '@aws-cdk/aws-iam';
import * as iot from '@aws-cdk/aws-iot';
import * as sns from '@aws-cdk/aws-sns';
import { singletonTopicRuleRole } from './util';

/**
 * The allowd message formats
 */
export enum MessageFormats {
  /**
   * JSON topic message format
   */
  JSON = 'JSON',
  /**
   * RAW topic message format
   */
  RAW = 'RAW',
}
/**
 * Construction properties for a sns publish action.
 */
export interface SnsTopicProps {
  /**
   * (Optional) The message format of the message to publish. Accepted values
   * are "JSON" and "RAW". The default value of the attribute is "RAW". SNS uses
   * this setting to determine if the payload should be parsed and relevant
   * platform-specific bits of the payload should be extracted.
   *
   * https://docs.aws.amazon.com/sns/latest/dg/json-formats.html
   *
   * @default - MessageFormats.RAW
   */
  readonly messageFormat?: MessageFormats;
  /**
   * The IAM role that grants access.
   *
   * @default - a role will be created
   */
  readonly role?: iam.IRole;
}

/**
 * Publishes to a Topic
 */
export class SnsTopic implements iot.ITopicRuleAction {
  constructor(private readonly topic: sns.ITopic, private readonly props: SnsTopicProps = {}) {
  }

  public bind(_rule: iot.ITopicRule): iot.TopicRuleActionConfig {
    // Allow rule to publish to topic
    const grantable = this.props.role ? this.props.role : new iam.ServicePrincipal('iot.amazonaws.com');
    this.topic.grantPublish(grantable);

    const role = this.props.role ? this.props.role : singletonTopicRuleRole(_rule, []);

    return {
      sns: {
        messageFormat: this.props.messageFormat || MessageFormats.RAW,
        targetArn: this.topic.topicArn,
        roleArn: role.roleArn,
      },
    };
  }
}
