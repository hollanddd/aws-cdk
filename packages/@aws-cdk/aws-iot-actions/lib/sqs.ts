import * as iam from '@aws-cdk/aws-iam';
import * as iot from '@aws-cdk/aws-iot';
import * as sqs from '@aws-cdk/aws-sqs';
import { singletonTopicRuleRole } from './util';

/**
 * Construction properties for a sqs send message action.
 */
export interface SqsQueueProps {
  /**
   * Specifies whether to use Base64 encoding.
   *
   * @default - false
   */
  readonly useBase64?: boolean;
  /**
   * The IAM role that grants access.
   *
   * @default - a role will be created
   */
  readonly role?: iam.IRole;
}

/**
 * Publishes to a Queue
 */
export class SqsQueue implements iot.ITopicRuleAction {
  constructor(private readonly queue: sqs.IQueue, private readonly props: SqsQueueProps = {}) {
  }

  public bind(_rule: iot.ITopicRule): iot.TopicRuleActionConfig {
    // Allow rule to publish to topic
    const grantable = this.props.role ? this.props.role : new iam.ServicePrincipal('iot.amazonaws.com');
    this.queue.grantSendMessages(grantable);

    const role = this.props.role ? this.props.role : singletonTopicRuleRole(_rule, []);

    return {
      sqs: {
        useBase64: this.props.useBase64 || false,
        queueUrl: this.queue.queueUrl,
        roleArn: role.roleArn,
      },
    };
  }
}
