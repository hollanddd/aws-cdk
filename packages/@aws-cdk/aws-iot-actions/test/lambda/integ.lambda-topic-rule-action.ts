import * as iot from '@aws-cdk/aws-iot';
import * as lambda from '@aws-cdk/aws-lambda';

import * as cdk from '@aws-cdk/core';
import * as actions from '../../lib';

/**
 * Define a rule that triggers a Lambda function when data is received.
 *
 * Automatically creates invoke permission
 */
const app = new cdk.App();

const stack = new cdk.Stack(app, 'aws-cdk-lambda-topic-rule-action');

const func = new lambda.Function(stack, 'Function', {
  code: lambda.Code.fromInline('boom'),
  handler: 'index.handler',
  runtime: lambda.Runtime.NODEJS_12_X,
});

const rule = new iot.TopicRule(stack, 'MyIotTopicRule', {
  sql: 'SELECT * FROM \'topic/subtopic\'',
});

rule.addAction(new actions.LambdaFunction(func));

app.synth();
