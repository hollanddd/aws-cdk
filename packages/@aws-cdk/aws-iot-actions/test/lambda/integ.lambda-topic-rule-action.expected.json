{
  "Resources": {
    "FunctionServiceRole675BB04A": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "Function76856677": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "ZipFile": "boom"
        },
        "Role": {
          "Fn::GetAtt": [
            "FunctionServiceRole675BB04A",
            "Arn"
          ]
        },
        "Handler": "index.handler",
        "Runtime": "nodejs12.x"
      },
      "DependsOn": [
        "FunctionServiceRole675BB04A"
      ]
    },
    "FunctionAllowIotD2ECA9CD": {
      "Type": "AWS::Lambda::Permission",
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "Function76856677",
            "Arn"
          ]
        },
        "Principal": "iot.amazonaws.com",
        "SourceAccount": {
          "Ref": "AWS::AccountId"
        }
      }
    },
    "MyIotTopicRuleFE4E2C8B": {
      "Type": "AWS::IoT::TopicRule",
      "Properties": {
        "TopicRulePayload": {
          "Actions": [
            {
              "Lambda": {
                "FunctionArn": {
                  "Fn::GetAtt": [
                    "Function76856677",
                    "Arn"
                  ]
                }
              }
            }
          ],
          "AwsIotSqlVersion": "2015-10-08",
          "RuleDisabled": false,
          "Sql": "SELECT * FROM 'topic/subtopic'"
        }
      },
      "DependsOn": [
        "FunctionAllowIotD2ECA9CD"
      ]
    }
  }
}